### PR TITLE
Start "IREE Best Practices" document.

### PR DESCRIPTION
**Pull request by ScottTodd**
Trying to put some of our intuition and general IREE design goals down on [virtual] paper that we can point people at, inspired by pages like https://www.tensorflow.org/lite/performance/best_practices (for layout and general topics to cover). Some things in here like "track state within your model" are responses to existing practices for working with current-generation frameworks that we want people to untangle in their models.

I don't really have an upper limit on how long this document could be in mind. I'm starting this page on `docs/developers/`, but we could move it to `docs/website/` to have it accessible on https://google.github.io/iree/ if it is useful to enough people.

### CODE REVIEW

File: docs/developers/best_practices.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+* keep compute devices saturated with work through pipelining
+* use dense math where possible, particular for inner loop bodies
+* limit synchronization points between devices like CPUs and GPUs
+* profile early and often, using the right tools for each level of granularity
+
+## Practices for model authoring
+
+### Track state within your model when possible
+
+If your model is stateful, prefer to store that state directly within your
```
Comment (benvanik) [inline docs/developers/best_practices.md:23]: grammar nit: remove all commas in this paragraph - they aren't needed with compound predicates

File: docs/developers/best_practices.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+working on your own applications.
+
+## Introduction
+
+Common themes include:
+
+* give the compiler as much information as possible
+* give the compiler opportunities to batch work together or defer computation
+* keep compute devices saturated with work through pipelining
+* use dense math where possible, particular for inner loop bodies
```
Comment (benvanik) [inline docs/developers/best_practices.md:15]: particularly*

**Comment by benvanik** [comment]
> we could move it to `docs/website/` to have it accessible on https://google.github.io/iree/ if it is useful to enough people.

I think this level of documentation is best on the website as it's for people developing models to pass to IREE vs. IREE developer documentation (how to develop the compiler/runtime, etc).

**Comment by ScottTodd** [comment]
> > we could move it to `docs/website/` to have it accessible on https://google.github.io/iree/ if it is useful to enough people.
> 
> I think this level of documentation is best on the website as it's for people developing models to pass to IREE vs. IREE developer documentation (how to develop the compiler/runtime, etc).

SG. I'll look at our options for navigation sections again (https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#sticky-navigation-tabs), since we have a lot of pages now (new blog section is landing, this page, maybe emscripten / wasm pages in the future, etc.)

**Comment by benvanik** [comment]
Having it under ML frameworks seems fine for now ("General best practices" or whatnot), but yeah I feel like a section that covers high-level user scenarios would be good (we don't have anything like that today)

**Comment by antiagainst** [comment]
+1. Also for this doc, we can have a specific "intended audience" or "scope" section describing what's this for. It can be a bit confusing if I just jump into it following a link and didn't study IREE's website structure well before.

**Comment by ScottTodd** [comment]
> +1. Also for this doc, we can have a specific "intended audience" or "scope" section describing what's this for. It can be a bit confusing if I just jump into it following a link and didn't study IREE's website structure well before.

Does 

> This page contains a list of best practices for getting the most out of IREE,
spanning model authoring, ahead-of-time compilation, and runtime use. Treat
these as a collection of ideas to consider or areas to start benchmarking when
working on your own applications.

not already cover that?

**Comment by benvanik** [comment]
I think it's fine - ship it!

**Review by benvanik — APPROVED** [review]

**Pull request merged**